indexprovider/wrapper.go

```diff
@@ -36,55 +40,107 @@ var shardRegMarker = ".boost-shard-registration-complete"
+func NewWrapper(cfg *config.Boost) func(lc fx.Lifecycle, h host.Host, r repo.LockedRepo, dealsDB *db.DealsDB,
 	legacyProv lotus_storagemarket.StorageProvider, prov provider.Interface, dagStore *dagstore.Wrapper,
-	meshCreator idxprov.MeshCreator) *Wrapper {
+	meshCreator idxprov.MeshCreator) (*Wrapper, error) {
 
-	return func(lc fx.Lifecycle, r repo.LockedRepo, dealsDB *db.DealsDB,
+	return func(lc fx.Lifecycle, h host.Host, r repo.LockedRepo, dealsDB *db.DealsDB,
 		legacyProv lotus_storagemarket.StorageProvider, prov provider.Interface, dagStore *dagstore.Wrapper,
-		meshCreator idxprov.MeshCreator) *Wrapper {
+		meshCreator idxprov.MeshCreator) (*Wrapper, error) {
 		if cfg.DAGStore.RootDir == "" {
 			cfg.DAGStore.RootDir = filepath.Join(r.Path(), defaultDagStoreDir)
 		}
 
 		_, isDisabled := prov.(*DisabledIndexProvider)
+
+		var eps []xproviders.Info
+
+		// add bitswap extended provider as needed
+		if cfg.Dealmaking.BitswapPeerID != "" {
+			meta := metadata.Default.New(metadata.Bitswap{})
+			mbytes, err := meta.MarshalBinary()
+			if err != nil {
+				return nil, err
+			}
+			if len(cfg.Dealmaking.BitswapPublicAddresses) > 0 {
+				keyFile, err := os.ReadFile(cfg.Dealmaking.BitswapPrivKeyFile)
+				if err != nil {
+					return nil, err
+				}
+				privKey, err := crypto.UnmarshalPrivateKey(keyFile)
+				if err != nil {
+					return nil, err
+				}
+				eps = append(eps, xproviders.Info{
+					ID:       cfg.Dealmaking.BitswapPeerID,
+					Addrs:    cfg.Dealmaking.BitswapPublicAddresses,
+					Priv:     privKey,
+					Metadata: mbytes,
+				})
+			} else {
+				eps = append(eps, xproviders.NewInfo(h.ID(), h.Peerstore().PrivKey(h.ID()), mbytes, h.Addrs()))
+			}
+		}
+
 		w := &Wrapper{
-			dealsDB:        dealsDB,
-			legacyProv:     legacyProv,
-			prov:           prov,
-			dagStore:       dagStore,
-			meshCreator:    meshCreator,
-			cfg:            cfg.DAGStore,
-			bitswapEnabled: cfg.Dealmaking.BitswapPeerID != "",
-			enabled:        !isDisabled,
+			dealsDB:           dealsDB,
+			legacyProv:        legacyProv,
+			prov:              prov,
+			dagStore:          dagStore,
+			meshCreator:       meshCreator,
+			cfg:               cfg.DAGStore,
+			extendedProviders: eps,
+
+			enabled: !isDisabled,
 		}
 		// announce all deals on startup in case of a config change
 		lc.Append(fx.Hook{
 			OnStart: func(ctx context.Context) error {
 				go func() {
-					_ = w.IndexerAnnounceAllDeals(ctx)
+					_ = w.AnnounceExtendedProviders(ctx)
 				}()
 				return nil
 			},
 		})
-		return w
+		return w, nil
 	}
 }
 
 func (w *Wrapper) Enabled() bool {
 	return w.enabled
 }
 
+func (w *Wrapper) AnnounceExtendedProviders(ctx context.Context) error {
+
+	if len(w.extendedProviders) == 0 {
+		return nil
+	}
+	adBuilder := xproviders.NewAdBuilder(w.h.ID(), w.h.Peerstore().PrivKey(w.h.ID()), w.h.Addrs())
+	adBuilder.WithExtendedProviders(w.extendedProviders...)
```

Comment:
If you want a bitswap endpoint to also be added for the main provider, don't forget to call `.WithMetadata(...)` on the builder.